src/collection_functions.php

```diff
@@ -61,113 +65,117 @@ function size($collection)
 /**
  * Returns a non-lazy collection with items from $collection in reversed order.
  *
- * @param array|Traversable $collection
- * @return Collection
+ * @template TKey
+ * @template TVal
+ * @param iterable<TKey, TVal> $collection
+ * @return iterable<TKey, TVal>
  */
-function reverse($collection)
+function reverse(iterable $collection): iterable
 {
-    $generatorFactory = function () use ($collection) {
+    $factory = function () use ($collection): Generator {
```

Comment:
Sometimes it is used  `Generator` as return type, other times like in the next change set (function values) it is used `iterable`. I would suggest use always `iterable`, for consistency unless Generator is required in the inner function, in which case i would then change all of them to Generator.